modmail/extensions/configuration_manager.py

```diff
@@ -0,0 +1,221 @@
+import logging
+import operator
+import string
+import types
+import typing
+
+import attr
+import attr._make
+import marshmallow
+from discord.ext import commands
+from discord.ext.commands import Context
+from discord.ext.commands import converter as commands_converter
+
+from modmail import config
+from modmail.bot import ModmailBot
+from modmail.log import ModmailLogger
+from modmail.utils.cogs import ExtMetadata, ModmailCog
+from modmail.utils.pagination import ButtonPaginator
+
+
+EXT_METADATA = ExtMetadata()
+
+logger: ModmailLogger = logging.getLogger(__name__)
+
+KeyT = str
+
+
+@attr.mutable
+class ConfOptions:
+    """Configuration attribute class."""
+
+    default: str
+    name: str
+    description: str
+    canonical_name: str
+    extended_description: str
+    hidden: bool
+
+    _type: type
+
+    metadata: dict
+
+    modmail_metadata: config.ConfigMetadata
+
+    _field: attr.Attribute = None
+    nested: str = None
+    frozen: bool = False
+
+    @classmethod
+    def from_field(cls, field: attr.Attribute, *, frozen: bool = False, nested: str = None):
+        """Create a ConfOptions from a attr.Attribute."""
+        kw = {}
+        kw["default"] = field.default if field.default is not marshmallow.missing else None
+        kw["name"] = field.name
+        kw["type"] = field.type
+
+        kw["metadata"] = field.metadata
+        kw["field"] = field
+
+        kw["frozen"] = field.on_setattr is attr.setters.frozen or frozen
+
+        meta: config.ConfigMetadata = field.metadata[config.METADATA_TABLE]
+        kw[config.METADATA_TABLE] = meta
+        kw["description"] = meta.description
+        kw["canonical_name"] = meta.canonical_name
+        kw["extended_description"] = meta.extended_description
+        kw["hidden"] = meta.hidden
+
+        if nested is not None:
+            kw["nested"] = nested
+
+        return cls(**kw)
```

Comment:
Any reason to be building a dict rather than just pass everything as kw args?

Reply:
Not really, there's just a lot of fields and some logic here.